buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java

```diff
@@ -98,6 +164,11 @@ public void transform() throws IOException {
         }
     }
 
+    @Input
+    public List<String> getTransformations() {
```

Comment:
I'm now starting to realize that this may be insufficient as we aren't capturing any of the _configuration_ of these transformations. For example, if I have `replaceMatch('foo', 'bar')` and then change it to `replaceMatch('foo', 'baz'` we'll get a false up-to-date because the only thing we track here is that we have configured a `ReplaceMatch` transform, but we're ignoring the fact that it's configured differently. This is my oversight, I didn't catch the fact that we pass config when instantiating these things. I have another proposed solution :wink: 

1. We revert to have `getTransformations()` return the raw set of `RestTestTransform` again.
2. Instead of using `@Input` we annotation it with `@Nested` which tells Gradle, "go look for any input annotations on each of these things, rather than treat the thing itself as an input".
3. For each `RestTestTransform` implementation we add getters for each of the constructor arguments, and annotate those with `@Input`.
4. Since some transform implementations might have ambiguous inputs, we still want to track what "kind" of transform as part of the input as well so make sure `RestTestTransform extends Named` and each transform should implement `getName()` by returning the FQN of the implementation class.

This should mean if I change any of the types of transforms, OR the constructor arguments I pass to them, the task will be out of date.

@breskeby please sanity check my statements above.

Reply:
Alternatively, The transforms are pretty easy to codify as a string, I could implement toString() (or some type of canonical string representation to avoid toString() method such a `asString()`)  and just map this method to the that method instead of the className. Same outcome (I think) but avoids adding Gradle specific annotations on the transformation code, and imo sounds abit simpler.

Reply:
I'd prefer we use proper `@Input` annotations instead of flattening this thing into a string. One reason for this is that when trying to troubleshoot build cache or up-to-date issues, a build scan comparison will tell us exactly which property has changed rather than just "yeah, some transform is different".

Reply:
> Same outcome (I think) but avoids adding Gradle specific annotations on the transformation code, and imo sounds abit simpler.

Either way these transforms are coupled to Gradle in that in the `toString()` scenario we have an expectation that that implementation captures everything relevant to input snapshotting. Using annotations just expresses this contract more explicitly.

Reply:
done on [`db82fd8` (#68338)](https://github.com/elastic/elasticsearch/pull/68338/commits/db82fd8671d4ce610f974838dbb59e25ee06b78d) and [`2b08fcb` (#68338)](https://github.com/elastic/elasticsearch/pull/68338/commits/2b08fcb06954e6f70716b3ea1e0038b00e17ba2e)...that was easier then I thought it would be, thanks for the very specific instructions. FWIW, I wrote the test first and it failed, then added the nested/input and it passed.